File: frameworks/node/koa/UnitTests/compressed.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";

import { client, expected } from "./app.ts";

const get = (size: string, encoding: string) => client.get(`/compressed/${size}`).set({ "accept-encoding": encoding, "cache-control": "no-cache" });

// superagent unzips a gzipped answer before it is read, so the encoding is read off the header.

// rb:test compressed.gzip_large
test("compressed.gzip_large: a large body is gzipped when asked", async () => {
  const response = await get("large", "gzip");

  assert.equal(response.headers["content-encoding"], "gzip");
  assert.deepEqual(response.body, expected.json("items.large.json"));
});

// rb:test compressed.gzip_small
test("compressed.gzip_small: a body under the threshold goes out as it is", async () => {
  const response = await get("small", "gzip");

  assert.equal(response.headers["content-encoding"], undefined);
  assert.deepEqual(response.body, expected.json("items.small.json"));
});

// rb:test compressed.identity_small,compressed.identity_large
for (const [id, size] of [["compressed.identity_small", "small"], ["compressed.identity_large", "large"]] as const) {
  test(`${id}: identity is answered as it is, and the handler runs every time`, async () => {
    const first = await get(size, "identity");
    const second = await get(size, "identity");

    assert.equal(second.headers["content-encoding"], undefined);
    assert.deepEqual(second.body, expected.json(`items.${size}.json`));
    assert.ok(Number(second.headers["x-rb-serial"]) > Number(first.headers["x-rb-serial"]));
  });
}

test("a route outside the family is not compressed", async () => {
  const response = await client.get("/json/large").set("accept-encoding", "gzip");

  assert.equal(response.headers["content-encoding"], undefined);
});

File: frameworks/go/huma/UnitTests/compressed_test.go
package unittests

import (
	"bytes"
	"compress/gzip"
	"io"
	"net/http"
	"reflect"
	"testing"
)

// rb:test compressed.gzip_small,compressed.gzip_large
func TestGzipIsAnsweredWhenTheRequestAsksForIt(t *testing.T) {
	for _, row := range []struct{ id, size string }{
		{"compressed.gzip_small", "small"},
		{"compressed.gzip_large", "large"},
	} {
		t.Run(row.id, func(t *testing.T) {
			before := serial(t, client(t).Get("/compressed/"+row.size, "Accept-Encoding: identity"))

			response := client(t).Get("/compressed/"+row.size, "Accept-Encoding: gzip", "Cache-Control: no-cache")

			assertStatus(t, response, http.StatusOK)
			assertHeader(t, response, "Content-Encoding", "gzip")
			reader, err := gzip.NewReader(bytes.NewReader(response.Body.Bytes()))
			if err != nil {
				t.Fatal(err)
			}
			plain, err := io.ReadAll(reader)
			if err != nil {
				t.Fatal(err)
			}
			if got, want := parsed(t, plain), payload(t, "items."+row.size+".json"); !reflect.DeepEqual(got, want) {
				t.Fatal("the gzipped body is not the payload")
			}
			if serial(t, response) <= before {
				t.Fatal("x-rb-serial did not advance")
			}
		})
	}
}

// rb:test compressed.identity_small,compressed.identity_large
func TestIdentityIsAnsweredUncompressed(t *testing.T) {
	for _, row := range []struct{ id, size string }{
		{"compressed.identity_small", "small"},
		{"compressed.identity_large", "large"},
	} {
		t.Run(row.id, func(t *testing.T) {
			response := client(t).Get("/compressed/"+row.size, "Accept-Encoding: identity", "Cache-Control: no-cache")

			assertNoHeader(t, response, "Content-Encoding")
			assertOK(t, payload(t, "items."+row.size+".json"), response)
		})
	}
}

func TestAWeightOfZeroRefusesGzip(t *testing.T) {
	response := client(t).Get("/compressed/small", "Accept-Encoding: gzip;q=0, identity")

	assertNoHeader(t, response, "Content-Encoding")
	assertOK(t, payload(t, "items.small.json"), response)
}

func TestEitherAnswerSaysItVariesByAcceptEncoding(t *testing.T) {
	for _, accept := range []string{"gzip", "identity"} {
		response := client(t).Get("/compressed/small", "Accept-Encoding: "+accept)

		assertHeader(t, response, "Vary", "Accept-Encoding")
	}
}
